cxx-qt-build/src/lib.rs

```diff
@@ -130,126 +129,97 @@ impl GeneratedCpp {
             "{}/{}.cxx.cpp",
-            source_directory_path.display(),
+            cpp_directory.display(),
             self.file_ident
         ));
         let mut cpp = File::create(&cpp_path).expect("Could not create cxx source file");
         cpp.write_all(&self.cxx.implementation)
             .expect("Could not write cxx source file");
-        written_files.push(cpp_path);
+        cpp_file_paths.plain_cpp = cpp_path;
 
-        written_files
+        cpp_file_paths
     }
 }
 
 /// Generate C++ files from a given list of Rust files, returning the generated paths
-fn write_cxx_generated_files_for_cargo(rs_source: &[&'static str]) -> Vec<PathBuf> {
+fn write_cxx_generated_files_for_cargo(
+    rs_source: &[&'static str],
+    header_dir: &impl AsRef<Path>,
+) -> Vec<GeneratedCppFilePaths> {
     let manifest_dir = manifest_dir();
-    let directory = format!("{}/target/cxx-qt-gen", manifest_dir);
-    std::fs::create_dir_all(&directory).expect("Could not create cxx-qt code generation directory");
-
-    let mut cpp_files = Vec::new();
+    let cpp_directory = format!("{}/cxx-qt-gen/src", env::var("OUT_DIR").unwrap());
 
+    let mut generated_file_paths: Vec<GeneratedCppFilePaths> = Vec::new();
     for rs_path in rs_source {
         let path = format!("{}/{}", manifest_dir, rs_path);
         println!("cargo:rerun-if-changed={}", path);
 
         let generated_code = GeneratedCpp::new(&path);
-        cpp_files.append(&mut generated_code.write_to_directory(&directory));
-    }
-
-    cpp_files
-}
-
-/// Write the list of C++ paths to the file
-fn write_cpp_sources_list(paths: &[PathBuf]) {
-    let manifest_dir = manifest_dir();
-
-    let path = format!("{}/target/cxx-qt-gen", manifest_dir);
-    std::fs::create_dir_all(path).expect("Could not create target dir");
-
-    let path = format!("{}/target/cxx-qt-gen/cpp_sources.txt", manifest_dir);
-    let mut file = File::create(&path).expect("Could not create cpp_sources file");
-
-    for path in paths {
-        writeln!(file, "{}", path.display()).unwrap();
+        generated_file_paths.push(generated_code.write_to_directory(&cpp_directory, header_dir));
     }
-}
-
-/// Write out the static header file for both the cxx
-fn write_cxx_static_header() {
-    let manifest_dir = manifest_dir();
 
-    let path = format!("{}/target/cxx-qt-gen/statics/rust", manifest_dir);
-    std::fs::create_dir_all(&path).expect("Could not create static header dir");
-
-    let h_path = format!("{}/cxx.h", path);
-    let mut header = File::create(&h_path).expect("Could not create cxx.h");
-    write!(header, "{}", cxx_gen::HEADER).expect("Could not write cxx.h");
+    generated_file_paths
 }
 
 /// Describes a cxx Qt builder which helps parse and generate sources for cxx-qt
 #[derive(Default)]
 pub struct CxxQtBuilder {
     cpp_format: Option<ClangFormatStyle>,
     rust_sources: Vec<&'static str>,
-    qt_enabled: bool,
+    pub cc_builder: cc::Build,
```

Comment:
Should this be a public field? or a getter ? 

But hmm looking further down the include() calls break the builder pattern. Why not have `include(mut self, ...)` as a method on `CxxQtBuilder` so that you can do

```rust
CxxQtBuilder::new()
        .cpp_format(ClangFormatStyle::Mozilla)
        .file("src/lib.rs")
        .include("include")
        .build();
```

instead of 

```rust
    let mut cxxqt_builder = CxxQtBuilder::new()
        .cpp_format(ClangFormatStyle::Mozilla)
        .file("src/lib.rs");
    cxxqt_builder.cc_builder.include("include");
    cxxqt_builder.build();
```

Reply:
We could make wrapper functions for the cc::Builder, but then it would be up to us to track any API changes to cc::Builder. So there is a trade off between maintenance burden for us and keeping strictly to a builder pattern API. I'd prefer the former.

Reply:
OK what about a closure in the builder pattern to access and mutate the inner cc_builder, like this ? This then keeps the builder pattern, but exposes the cc_builder object so we don't have a larger maintenance burden ?

```rust
CxxQtBuilder::new()
        .cpp_format(ClangFormatStyle::Mozilla)
        .file("src/lib.rs")
        .cc_builder(|cc| cc.include("include"))
        .build();
```

Reply:
clever. I like it :)

Reply:
Implemented. Now the diffs for the build.rs scripts are trivial :)